opentelemetry-sdk/src/opentelemetry/sdk/metrics/export/exemplars.py

```diff
@@ -0,0 +1,349 @@
+        """The span ID of the context when the exemplar was recorded"""
+        return self._span_id
+
+    @property
+    def trace_id(self):
+        """The trace ID of the context when the exemplar was recorded"""
+        return self._trace_id
+
+    @property
+    def dropped_labels(self):
+        """Labels that were dropped by the aggregator but still passed by the user"""
+        return self._dropped_labels
+
+    @property
+    def sample_count(self):
+        """For statistical exemplars, how many measurements a single exemplar represents"""
+        return self._sample_count
+
+    def set_sample_count(self, count):
+        self._sample_count = count
+
+
+class ExemplarSampler:
+    """
+        Abstract class to sample exemplars through a stream of incoming measurements
+    """
+
+    def __init__(self, k, statistical=False):
+        self._k = k
+        self._statistical = statistical
+        self._sample_set = list()
+
+    @abc.abstractmethod
+    def sample(self, exemplar, **kwargs):
+        """
+        Given an exemplar, determine if it should be sampled or not
+        """
+
+    @property
+    @abc.abstractmethod
+    def sample_set(self):
+        """
+        Return the list of exemplars that have been sampled
+        """
+
+    @abc.abstractmethod
+    def merge(self, set1, set2):
+        """
+        Given two lists of sampled exemplars, merge them while maintaining the invariants specified by this sampler
+        """
+
+    @abc.abstractmethod
+    def reset(self):
+        """
+        Reset the sampler
+        """
+
+
+class RandomExemplarSampler(ExemplarSampler):
+    """
+        Randomly sample a set of k exemplars from a stream. Each measurement in the stream
+        will have an equal chance of being sampled.
+
+        If RandomExemplarSampler` is specified to be statistical, it will add a sample_count to every exemplar it records.
+        This value will be equal to the number of measurements recorded per every exemplar measured - all exemplars will have the same sample_count value.
+    """
+
+    def __init__(self, k, statistical=False):
+        super().__init__(k, statistical=statistical)
+        self.rand_count = 0
+
+    def sample(self, exemplar, **kwargs):
+        self.rand_count += 1
+
+        if len(self.sample_set) < self._k:
+            self.sample_set.append(exemplar)
+            return
+
+        replace_index = random.randint(0, self.rand_count - 1)
+
+        if replace_index < self._k:
+            self.sample_set[replace_index] = exemplar
+
+    def merge(self, set1, set2):
+        combined = set1 + set2
+        if len(combined) <= self._k:
+            return combined
+        return random.sample(combined, self._k)
+
+    @property
+    def sample_set(self):
+        if self._statistical:
+            for exemplar in self._sample_set:
+                exemplar.set_sample_count(
+                    self.rand_count / len(self._sample_set)
+                )
```

Comment:
Could this be done by giving the exemplar a reference somehow so it could compute this on its own? Not so sure how that would work with merge

Reply:
I think it's fine like this? referencing the sampler from the exemplar seems worse..

Reply:
I'm not sure of all the ways this property and the exemplar are used outside of this class, but my point is the `exemplar.sample_count` will be stale once more samples come in, until this property is called again

```py
some_exemplar = sampler.sample_set[0]

# prints 150
print(some_exemplar.sample_count)

# ...
# sampler receives more samples in another thread

# still prints 150
print(some_exemplar.sample_count)

sampler.sample_set
# now adjusted
print(some_exemplar.sample_count)
```

Reply:
exemplars shouldn't be used until they are exported, so I think it's fine as is.